Migrate to Maps SDK v11

eugenes78 opened the pull request:
<!-- ⚠️ TEMPLATE ⚠️ -->
<!-- Template for GitHub PR descriptions. Use it as a guide on how to describe your work. Feel free to remove any section when you're opening a PR if you think it does not apply for your committed changes. -->

### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->


### Screenshots or Gifs
<!-- 
Include media files to provide additional context. It's REALLY useful for UI related PRs (e.g. ![screenshot gif](link))

Please note that you can change image width/height with one of the following ways:
- For specifying size in percentage:
<img src="https://user-images.githubusercontent.com/4005589/120349671-f86c4f80-c306-11eb-8c71-b903666c2bc.png" width=50% height=50%>
- For specifying size in pixels:
<img src="https://user-images.githubusercontent.com/4005589/120349671-f86c4f80-c306-11eb-8c71-b903666c2bc.png" width="200" height="200">
-->


### Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR (where applicable)
- [x] I have run tests and automatic checks locally
- [ ] I have run `pitest` check locally and checked that the coverage increased (or didn't change) or decreased insignificantly
- [ ] I have made corresponding changes to the documentation (where applicable)
- [x] I have grouped commits logically or I promise to squash my commits before merge

eugenes78 commented on MapboxSearch/autofill/build.gradle:
```diff
@@ -84,7 +84,7 @@ dependencies {
     testImplementation project(":common-tests")
 
-    androidTestImplementation "com.mapbox.common:okhttp:$common_sdk_version"
+    androidTestImplementation "com.squareup.okhttp3:okhttp:4.12.0"
```
@DzmitryFomchyn @ahmedaly16 
I guess we need to republish this package, am I wrong here?

DzmitryFomchyn replied:
No, we can't republish this because we don't control it, it's another team responsibility.

@eugenes78 try 

`okhttp_mock_version = '1.5.0'`
...
`testImplementation "com.github.gmazzo:okhttp-mock:$okhttp_mock_version"`

eugenes78 replied:
Resolved through a new variable called `okhttp_version`

DzmitryFomchyn commented on MapboxSearch/base/build.gradle:
```diff
@@ -68,7 +68,7 @@ dependencies {
 
     // Explicit dependency for http service implementation, should be declared on the Common SDK side
-    implementation "com.mapbox.common:okhttp:$common_sdk_version"
+    implementation "com.squareup.okhttp3:okhttp:4.12.0"
```
This can be removed at all.

eugenes78 replied:
Done

DzmitryFomchyn commented on MapboxSearch/base/build.gradle:
```diff
@@ -88,7 +88,7 @@ dependencies {
     testImplementation project(":common-tests")
 
-    androidTestImplementation "com.mapbox.common:okhttp:$common_sdk_version"
+    androidTestImplementation "com.squareup.okhttp3:okhttp:4.12.0"
```
It's better to extract `4.12.0` to constants in `dependencies.gradle`

eugenes78 replied:
It requires introducing a new file, I left it the way it is right now to minimize changes (there are a lot of them already).

DzmitryFomchyn commented on MapboxSearch/gradle/versions.gradle:
```diff
@@ -1,7 +1,7 @@
     gradle_plugin = '7.3.1'
-    kotlin_version = '1.5.31'
-    kotlin_coroutines_version = '1.5.2'
+    kotlin_version = '1.9.21'
```
It's better not to bump Kotlin version to latest because customers with older version of Kotlin and who can't bump it, won't be able to use the Search SDK.

Maps v11 is using `1.7.20`, it's better to use this version for search as well.

DzmitryFomchyn commented on DemoApps/SearchWithMaps/app/build.gradle:
```diff
@@ -41,7 +41,7 @@ android {
 dependencies {
     implementation "com.mapbox.search:mapbox-search-android-ui:1.0.0-rc.5"
-    implementation "com.mapbox.maps:android:10.14.0-beta.1"
+    implementation "com.mapbox.maps:android:11.0.0"
```
This change is not needed. You can bump it only together with `com.mapbox.search:mapbox-search-android-ui` dependency.

Note that `DemoApps` is a folder with external demo apps, its sources are not part of the SDK

eugenes78 replied:
Reverted

DzmitryFomchyn commented on MapboxSearch/ui/src/main/java/com/mapbox/search/ui/utils/extenstion/View.kt:
```diff
@@ -12,7 +12,11 @@ internal fun View.showKeyboard() {
 }
 
-internal fun View.hideKeyboard() {
+/**
```
Why do you need this change? When you develop SDK it's better to keep everything non-public, until proven otherwise.

eugenes78 replied:
I made it internal again. It was an attempt to avoid some code duplication.

DzmitryFomchyn commented on MapboxSearch/ui/src/main/java/com/mapbox/search/ui/utils/extenstion/Context.kt:
```diff
@@ -72,3 +83,51 @@ internal fun Context.resolveAttr(@AttrRes resId: Int): Int? {
     }
 }
+
+/**
```
Why do you need these changes? When you develop SDK it's better to keep everything non-public, until proven otherwise.

eugenes78 replied:
Because it doesn't work otherwise:
```
> Task :sample:compileDebugKotlin FAILED
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/AddressAutofillUiActivity.kt: (29, 46): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/AddressAutofillUiActivity.kt: (175, 21): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/AddressAutofillUiActivity.kt: (183, 17): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/AddressAutofillUiActivity.kt: (194, 17): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/DiscoverActivity.kt: (36, 46): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/DiscoverActivity.kt: (113, 25): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/DiscoverActivity.kt: (132, 21): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/PlaceAutocompleteUiActivity.kt: (37, 46): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/PlaceAutocompleteUiActivity.kt: (209, 21): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/PlaceAutocompleteUiActivity.kt: (215, 17): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
e: .../MapboxSearch/sample/src/main/java/com/mapbox/search/sample/PlaceAutocompleteUiActivity.kt: (232, 17): Cannot access 'showToast': it is internal in 'com.mapbox.search.ui.utils.extenstion'
```

DzmitryFomchyn replied:
If these functions are used only in sample module, it's better to move them there. I think these utilities have to be internal if they are really needed in ui module.

Api reference should be fixed anyway.

eugenes78 replied:
Made them internal again

DzmitryFomchyn commented on MapboxSearch/sdk/src/main/java/com/mapbox/search/SearchEngineSettings.kt:
```diff
@@ -22,7 +21,7 @@ public class SearchEngineSettings @JvmOverloads constructor(
      * [Manifest.permission.ACCESS_FINE_LOCATION] to work properly.
      */
-    public val locationEngine: LocationEngine = defaultLocationEngine(),
+    public val locationProvider: DeviceLocationProvider = defaultDeviceLocationProvider(),
```
It's better to reference base `com.mapbox.common.location.LocationProvider` interface if you don't need anything specific from `DeviceLocationProvider` in your implementation. Otherwise, everyone who wants to provide own Location Provider implementation will have to deal with unused methods.

This comment is applicable to all the classes that accept location provider as a parameter.

eugenes78 replied:
Done

DzmitryFomchyn commented on MapboxSearch/sdk/src/main/java/com/mapbox/search/SearchEngineSettings.kt:
```diff
@@ -22,7 +21,7 @@ public class SearchEngineSettings @JvmOverloads constructor(
      * Note that this class requires [Manifest.permission.ACCESS_COARSE_LOCATION] or
      * [Manifest.permission.ACCESS_FINE_LOCATION] to work properly.
      */
```
API ref is also should be updated.

eugenes78 replied:
What do you mean by "API ref"?

DzmitryFomchyn commented on MapboxSearch/sdk/src/main/java/com/mapbox/search/SearchEngineSettings.kt:
SearchEngineSettings.accessToken should be removed. Common 24 brings centralised token managment

eugenes78 replied:
Done

DzmitryFomchyn commented:
@eugenes78 I can highlight a few points that can be applied to the whole PR:

1. Replace DeviceLocationProvider with LocationProvider. It's better to reference base com.mapbox.common.location.LocationProvider interface if you don't need anything specific from DeviceLocationProvider in your implementation. Otherwise, everyone who wants to provide own Location Provider implementation will have to deal with unused methods.
2. Common 24 brings centralised token managment. Token should be set in top level files (in case of this repo sample project, this should be activity). From other parts of project (SearchEngineSettings, OfflineSearchEngineSettings, etc.) token should be removed

Once you do that, this PR will be much closer to final state.

In addition, it makes sense to clean up PR. There are a lot of changes that look accidental and don’t bring anything. When you clean up the PR, there should be much less files to review. Some things to clean up: DemoApps folder, commented blocks of code, functions that has been made public without need

eugenes78 commented:
@DzmitryFomchyn 

1. I replaced DeviceLocationProvider with LocationProvider as you suggested.
2. I cleaned up token handling per your suggestion as well.

Now I'm working on minor issues and, more importantly, failing integration tests.

DzmitryFomchyn commented on DemoApps/AddressAutofill/app/src/main/java/com/mapbox/demo/autofill/SampleAppUtils.kt:
```diff
@@ -34,14 +34,26 @@ fun LocationEngine.lastKnownLocation(context: Context, callback: (Point?) -> Uni
     })
 }
 
+/**
```
These changes are not needed

eugenes78 replied:
I'm getting errors if I try to remove "docs":
```
Undocumented: com.mapbox.search.ui.extensions/userDistanceTo/com.mapbox.common.location.LocationProvider#com.mapbox.geojson.Point#kotlin.Function1[kotlin.Double?,kotlin.Unit]/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/showToast/android.content.Context#kotlin.Int/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/isPermissionGranted/android.content.Context#kotlin.String/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/bitmapFromDrawableRes/android.content.Context#kotlin.Int/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/dpToPx/#kotlin.Int/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/geoIntent/#com.mapbox.geojson.Point/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/shareIntent/#com.mapbox.search.ui.view.place.SearchPlace/ (androidJvm)
Undocumented: com.mapbox.search.ui.utils.extenstion/hideKeyboard/android.view.View#/ (androidJvm)
```

DzmitryFomchyn replied:
Do you get thsse error when you build project in `MapboxSearch` dir?

Note that `DemoApps` is an independent directory with standalone apps. Also note the difference in package names `com.mapbox.search.ui` - `com.mapbox.demo`

DzmitryFomchyn commented on MapboxSearch/autofill/build.gradle:
```diff
@@ -84,10 +84,10 @@ dependencies {
     testImplementation project(":common-tests")
 
-    androidTestImplementation "com.mapbox.common:okhttp:$common_sdk_version"
+    androidTestImplementation "com.squareup.okhttp3:okhttp:$okhttp3_version"
```
[nit] This is not needed if you define `com.squareup.okhttp3:mockwebserver`

eugenes78 replied:
Fixed

DzmitryFomchyn commented on MapboxSearch/autofill/src/androidTest/java/com/mapbox/search/autofill/AddressAutofillIntegrationTest.kt:
```diff
@@ -46,15 +45,12 @@ internal class AddressAutofillIntegrationTest {
         val engine = createAutofillSearchEngine(
             app = APP,
-            token = TEST_ACCESS_TOKEN,
-            url = mockServer.url("").toString(),
```
This test shouldn't pass if you don't override `url`.

IT turned out that integration tests are not run on CI. See this PR https://github.com/mapbox/mapbox-search-android/pull/182

eugenes78 replied:
Done

DzmitryFomchyn commented on MapboxSearch/autofill/src/androidTest/java/com/mapbox/search/autofill/AddressAutofillIntegrationTest.kt:
```diff
@@ -78,7 +74,6 @@ internal class AddressAutofillIntegrationTest {
 
         val url = requireNotNull(request.requestUrl)
         assertEqualsIgnoreCase("//autofill/v1/suggest/${query.query}", url.encodedPath)
-        assertEquals(TEST_ACCESS_TOKEN, url.queryParameter("access_token"))
```
Token still should be passed to backend. This check should remain here. The same for other checks in this file

eugenes78 replied:
I rolled back that change. Fixed.

DzmitryFomchyn commented:
@eugenes78 it turned out that some integration tests for some of modules are not run on CI, please rebase once this PR merges https://github.com/mapbox/mapbox-search-android/pull/182

Note that you can run integration tests locally from IDE, or run all the checks locally using https://github.com/mapbox/mapbox-search-android?tab=readme-ov-file#complex-checks-run

DzmitryFomchyn commented on MapboxSearch/base/src/main/java/com/mapbox/search/base/location/LocationUtils.kt:
```diff
@@ -1,16 +1,18 @@
+    LocationServiceFactory.getOrCreate()
+        .getDeviceLocationProvider(null)
+        .value
+        ?: throw Exception("Failed to get device location provider")
```
This will crash process if location provider is not available. It's better to make `defaultLocationProvider()` return nullable type and assume that location provider might not be available.

eugenes78 replied:
Fixed

DzmitryFomchyn commented on MapboxSearch/base/src/test/java/com/mapbox/search/base/location/LocationEngineAdapterTest.kt:
```diff
@@ -66,70 +68,34 @@ internal class LocationEngineAdapterTest {
-    }
-
-    @TestFactory
-    fun `Requests new location if initial location request failed`() = TestCase {
```
What about this test case?

eugenes78 replied:
`LocationEngineAdapter` now uses `LocationProvider`, and I found no alternative method to check for a failure. So, I didn't see how I could rework this test to make it work.

DzmitryFomchyn commented on MapboxSearch/discover/src/androidTest/java/com/mapbox/search/discover/DiscoverIntegrationTest.kt:
```diff
@@ -75,7 +74,6 @@ internal class DiscoverIntegrationTest {
 
         val url = requireNotNull(request.requestUrl)
         assertEqualsIgnoreCase("//search/v1/category/${query.canonicalName}", url.encodedPath)
-        assertEquals(TEST_ACCESS_TOKEN, url.queryParameter("access_token"))
```
We still should check if token is passed to backend.

eugenes78 replied:
Fixed

DzmitryFomchyn commented on MapboxSearch/gradlew:
```diff
@@ -169,4 +169,5 @@ if [ "$(uname)" = "Darwin" ] && [ "$HOME" = "$PWD" ]; then
   cd "$(dirname "$0")"
 fi
 
+echo "$JAVACMD" "$@"
```
Why do you need this change?  This file shouldn't be updated manually

eugenes78 replied:
This was to figure out the exact version of Java that we have in CI/CD. Rolled back now.

DzmitryFomchyn commented on MapboxSearch/sample/src/main/java/com/mapbox/search/sample/api/AddressAutofillKotlinExampleActivity.kt:
```diff
@@ -16,9 +15,7 @@ class AddressAutofillKotlinExampleActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
-        addressAutofill = AddressAutofill.create(
```
Do all of the examples work for you? These examples are part of public documentation, this would be great to tell the customers how to set token.

DzmitryFomchyn commented on MapboxSearch/ui/src/main/java/com/mapbox/search/ui/extensions/DeviceLocatioonProvider.kt:
```diff
@@ -0,0 +1,22 @@
+ * TODO
+ *
+ */
+public fun LocationProvider.userDistanceTo(destination: Point, callback: (Double?) -> Unit) {
```
I'm not sure whether this function should be public. I'd make it internal.

Anyway, api reference should be fixed.

DzmitryFomchyn commented on MapboxSearch/ui/src/main/java/com/mapbox/search/ui/utils/extenstion/Context.kt:
```diff
@@ -72,3 +83,27 @@ internal fun Context.resolveAttr(@AttrRes resId: Int): Int? {
     }
 }
+
+internal fun Context.showToast(@StringRes resId: Int): Unit = Toast.makeText(this, resId, Toast.LENGTH_LONG).show()
```
All these added functions should be unused. Aren't they?

eugenes78 replied:
Removed

DzmitryFomchyn commented on MapboxSearch/ui/src/main/java/com/mapbox/search/ui/extensions/DeviceLocatioonProvider.kt:
```diff
@@ -0,0 +1,22 @@
+import com.mapbox.search.common.DistanceCalculator
+
+/**
+ * TODO
```
Please review all TODOs before merge

eugenes78 replied:
Resolved by removing the public method

aokj4ck reviewed (approved):
[Moved and seconded](https://robertsrules.com/)
